openhcl/underhill_attestation/src/igvm_attest/key_release.rs

```diff
@@ -162,14 +168,52 @@ pub fn parse_response(
     Ok(wrapped_key)
 }
 
+/// Convert a potentially non UTF-8 byte array into a string with non UTF-8 characters represented
+/// as hexadecimal escape sequences.
+fn string_from_utf8_preserve_invalid_bytes<T: AsRef<[u8]>>(bytes: T) -> String {
```

Comment:
Should the function take `bytes` and `Utf8Error` as the argument instead? The `match` seems redundant

Reply:
The function is recursive so handling the case where the bytes are actually valid UTF-8 is useful

Reply:
got it

Reply:
Uhh, hold on a second - let me review this code.

I'm seeing a recursive function that allocates a new String on each run, and that's a real eyebrow raiser in the domain we work in :eyes:

Reply:
@daprilik in https://github.com/microsoft/openvmm/pull/583/commits/fad817a8ecbe36552083102d0646ed027e22a036 I followed clippy's advice and swapped the string allocations for a more efficient approach: https://rust-lang.github.io/rust-clippy/master/index.html#/format_collect.

This function is only intended to be used in the case of error handling so that may help relax performance requirements?

Reply:
I'll sprinkle a comment or two on the current implementation, though honestly, you might be better off taking a dependency on https://docs.rs/bstr/latest/bstr/, and letting it do the heavy lifting here.

Reply:
I'd rather just include the raw bytes in the error message than introduce an external dependency just for this reason so if this gets too complicated/expensive I'll toss the string building helper function

Reply:
I'm also wary of adding new dependencies, but I'll note that `bstr` is a solid lib, and one I don't personally have concerns taking a dependency on. Moreover, I can think of many other places in the codebase that might benefit from using `bstr` vs. having to pass around raw u8 / sanitized strs, so having `bstr` encroach into the codebase may not be the worst thing.

Reply:
Sounds good, I took another crack and this on my own and I'll check out how `bstr` could help